src/client/datascience/raw-kernel/rawKernel.ts

```
@@ -1,527 +1,273 @@
-            }
-
-            // Dispose of all our outstanding futures
-            this.futures.forEach((future) => {
-                future.dispose();
-            });
-            this.futures.clear();
-
-            this.isDisposed = true;
-        }
-    }
-    public shutdown(): Promise<void> {
-        throw new Error('Not yet implemented');
-    }
-    public getSpec(): Promise<Kernel.ISpecModel> {
-        throw new Error('Not yet implemented');
+    ): Kernel.IShellFuture<
+        KernelMessage.IShellMessage<T>,
+        KernelMessage.IShellMessage<KernelMessage.ShellMessageType>
+    > {
+        return this.realKernel.sendShellMessage(msg, expectReply, disposeOnDone);
     }
     public sendControlMessage<T extends KernelMessage.ControlMessageType>(
-        _msg: KernelMessage.IControlMessage<T>,
-        _expectReply?: boolean,
-        _disposeOnDone?: boolean
-    ): Kernel.IControlFuture<KernelMessage.IControlMessage<T>> {
-        throw new Error('Not yet implemented');
+        msg: KernelMessage.IControlMessage<T>,
+        expectReply?: boolean,
+        disposeOnDone?: boolean
+    ): Kernel.IControlFuture<
+        KernelMessage.IControlMessage<T>,
+        KernelMessage.IControlMessage<KernelMessage.ControlMessageType>
+    > {
+        return this.realKernel.sendControlMessage(msg, expectReply, disposeOnDone);
     }
     public reconnect(): Promise<void> {
-        throw new Error('Not yet implemented');
+        return this.realKernel.reconnect();
     }
-    public async interrupt(): Promise<void> {
-        await this.kernelInterrupter();
+    public interrupt(): Promise<void> {
+        // Send this directly to our kernel process. Don't send it through the real kernel. The
+        // real kernel will send a goofy API request to the websocket.
+        return this.kernelProcess.interrupt();
     }
     public restart(): Promise<void> {
-        throw new Error('Not yet implemented');
+        return this.realKernel.restart();
```

I don't think this is correct for raw kernel.
I think we should instead throw an exception (cuz we don't support this scenario).

---

Yeah it should probably throw an exception. This never gets called.

---
In reply to: [416045862](https://github.com/microsoft/vscode-python/pull/11444#discussion_r416045862) [](ancestors = 416045862)

---

Yeah, restart and change both happen at the session level that we own.